org-code-javabuilder/lib/src/main/java/org/code/javabuilder/JavaRunner.java

```diff
@@ -1,22 +1,63 @@
 package org.code.javabuilder;
 
-import java.util.Scanner;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URL;
+import java.net.URLClassLoader;
 
-/**
- * Currently, JavaRunner just runs a sample student code. We'll add more here when we execute
- * arbitrary student code.
- */
+/** The class that executes the student's code */
 public class JavaRunner {
-  public void runCode() {
-    // Sample program. In the future, student code will be executed here.
-    System.out.println("What's your name? ");
-    Scanner in = new Scanner(System.in);
-    String s = in.nextLine();
-    System.out.print("Hello " + s + "!");
-    System.out.print("How ");
-    System.out.print("are ");
-    System.out.print("you? ");
-    s = in.nextLine();
-    System.out.println("You said: " + s);
+  private final URL executableLocation;
+  private final ProjectFileManager fileManager;
+  private final OutputAdapter outputAdapter;
+
+  public JavaRunner(
+      URL executableLocation, ProjectFileManager fileManager, OutputAdapter outputAdapter) {
+    this.executableLocation = executableLocation;
+    this.fileManager = fileManager;
+    this.outputAdapter = outputAdapter;
+  }
+
+  /**
+   * Run the compiled user code.
+   *
+   * @throws UserFacingException When the user's code hits a runtime error or fails due to an
+   *     internal error.
+   * @throws InternalFacingException When we hit an internal error after the user's code has
```

Comment:
the UserFacingExceptions and InternalFacingExceptions make the code super understandable!

Reply:
🎉